src/System.Text.Json/src/System/Text/Json/JsonDocument.Parse.cs

```diff
@@ -0,0 +1,276 @@
+
+            var database = new CustomDb(utf8Json.Length);
+            var stack = new CustomStack(JsonReaderState.DefaultMaxDepth * StackRow.Size);
+
+            try
+            {
+                Parse(utf8JsonSpan, reader, ref database, ref stack);
+
+                if (database.Length == 0)
+                {
+                    throw new JsonReaderException("Cannot load the empty document", -1, -1);
+                }
+            }
+            catch
+            {
+                database.Dispose();
+                throw;
+            }
+            finally
+            {
+                stack.Dispose();
+            }
+
+            return new JsonDocument(utf8Json, database, extraRentedBytes);
+        }
+
+        private static ArraySegment<byte> ReadToEnd(Stream stream)
+        {
+            int written = 0;
+            byte[] rented = null;
+            ArrayPool<byte> pool = ArrayPool<byte>.Shared;
+
+            try
+            {
+                long expectedLength = 0;
+
+                if (stream.CanSeek)
+                {
+                    expectedLength = Math.Max(1L, stream.Length - stream.Position);
+                    rented = pool.Rent(checked((int)expectedLength));
+                }
+                else
+                {
+                    rented = pool.Rent(UnseekableStreamInitialRentSize);
+                }
+
+                int lastRead;
+
+                do
+                {
+                    if (expectedLength == 0 && rented.Length == written)
+                    {
+                        byte[] tmp = pool.Rent(rented.Length * 2);
+                        Buffer.BlockCopy(rented, 0, tmp, 0, rented.Length);
+                        rented.AsSpan().Clear();
+                        pool.Return(rented);
+                        rented = tmp;
+                    }
+
+                    lastRead = stream.Read(rented, written, rented.Length - written);
+                    written += lastRead;
+                } while (lastRead > 0);
+
+                return new ArraySegment<byte>(rented, 0, written);
+            }
+            catch
+            {
+                if (rented != null)
+                {
+                    rented.AsSpan(0, written).Clear();
+                    pool.Return(rented);
+                }
+
+                throw;
+            }
+        }
+
+        private static async Task<ArraySegment<byte>> ReadToEndAsync(
+            Stream stream,
+            CancellationToken cancellationToken)
+        {
+            int written = 0;
+            byte[] rented = null;
+            ArrayPool<byte> pool = ArrayPool<byte>.Shared;
+
+            try
+            {
+                long expectedLength = 0;
+
+                if (stream.CanSeek)
+                {
+                    expectedLength = Math.Max(1L, stream.Length - stream.Position);
+                    rented = pool.Rent(checked((int)expectedLength));
```

Comment:
The JIT has a much better chance of devirtualizing this if you just write out `ArrayPool<byte>.Shared.Rent` instead of caching it into a local.